### DISCUSSION TITLE
Using ntp to set the Pico's hardware rtc

### DISCUSSION THREAD

**Discussion by w9ran** [earlephilhower/arduino-pico#2345] [Q&A]
I have a Pico-W application that needs to start a task at a precise time.   I've used the provided ntp client example to read time from ntp servers, but I want to use ntp to set the hardware rtc in the Pico periodically to maintain accuracy,  and use then use the functions provided in the sdk to read the time and to enable and disable alarms.   I figured someone had already done this but haven't found any examples (using C++ and the Arduino core),  so just thought I'd ask if anyone might know of any?   It seems to be a matter of reorganizing the ntp data to comply with the hardware.rtc binary format so examples showing how to do that would be helpful.  Thanks!

**Comment by earlephilhower** [comment]
Look at `ctime` which returns H/M/S D/M/Y and more in conjunction with any timezone you've set.

Not really sure why you'd want to set the RTC.  Unless you put the Pico to sleep internal time is kept track of just fine.  The RTC does have uses for waking up from "low power" sleep mode (but on the original Pico low power really isn't all that low), but it keeps track in such a weird "1970s digital watch" format that not many things can actually use...

**Reply by w9ran** [reply]
Well it wouldn't be the first time I've tried to make something harder than it actually is!   I'm not going to use sleep mode,  I just want to schedule events in real time without having to continually ping an ntp server, but to correct for time drift due to clock crystal error.     When I searched the Pico C SDK documentation for "time" what came up was the hardware API "hardware.rtc" and the High Level API pico_time so assumed the rtc was what I needed.    

Thanks for your suggestion of "ctime" which looks to be just what I need - I just didn't know it was available in the core.    I've got your setClock example running in the Arduino IDE but was confused by its use of functions like "time_t" and the type "struct_tm" which I could not find in either the Pico C sdk or your Arduino core.    Now that I see the are from ctime library - how can the example work with no "#include time.h" in the setClock example?   I assume the answer is that ctime is part of the Arduino core?

But - since the ctime library seems to be available I just need to learn how to use it to do what I want.    Thanks!

**Reply by earlephilhower** [reply]
We support the most of the [standard POSIX library](https://sourceware.org/newlib/) , of which `ctime` is one small bit.  Headers are included in `Arduino.h` (or its sub-includes) for it because `time.h` is very useful.  `time_t` is just "seconds since Jan 1, 1970" while `struct tm` is just "seconds and nanoseconds since Jan 1,1970) and standard.
